include/event_manager/event_manager.h

```diff
@@ -0,0 +1,317 @@
+ * - cast_<i>%event_type</i> - Casts the event header that is provided
+ *                            as argument to an event of the given type.
+ *
+ * @param ename		   Name of the event.
+ * @param init_log_en	   Bool indicating if the event is logged
+ *                         by default.
+ * @param log_fn	   Function to stringify an event of this type.
+ * @param ev_info_struct   Data structure describing the event type.
+ */
+#define EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, ev_info_struct) \
+	_EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, ev_info_struct)
+
+
+/** Verify if an event ID is valid.
+ *
+ * The pointer to an event type structure is used as its ID. This macro
+ * validates that the provided pointer is within the range where event
+ * type structures are defined.
+ *
+ * @param id  ID.
+ */
+#define ASSERT_EVENT_ID(id) \
+	__ASSERT_NO_MSG((id >= __start_event_types) && (id < __stop_event_types))
+
+
+/** Submit an event to the Event Manager.
+ *
+ * @param eh  Pointer to the event header element in the event object.
+ */
+void _event_submit(struct event_header *eh);
+
+
+/** Submit an event.
+ *
+ * This helper macro simplifies the event submission.
+ *
+ * @param event  Pointer to the event object.
+ */
+#define EVENT_SUBMIT(event) _event_submit(&event->header)
+
+
+/** Initialize the Event Manager.
+ *
+ * @retval 0 If the operation was successful.
+ */
+int event_manager_init(void);
+
+
+/** Trace event execution.
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
+ **/
+void event_manager_trace_event_execution(const struct event_header *eh,
+				  bool is_start);
+
+
+/** Trace event submission.
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
+ **/
+void event_manager_trace_event_submission(const struct event_header *eh,
+				const void *trace_info);
+
+
+/** Initialize tracing in the Event Manager.
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is no-operation.
+ * It is annotated as weak and is meant to be overridden by layer
+ * adding support for profiling mechanism.
+ **/
+int event_manager_trace_event_init(void);
+
+/** Allocate event.
+ * The behavior of this function depends on the actual implementation.
+ * The default implementation of this function is same as k_malloc.
+ * It is annotated as weak and can be overridden by user.
+ *
+ * @param size  Amount of memory requested (in bytes).
+ * @retval Address of the allocated memory if successful, otherwise NULL.
+ **/
+void *event_manager_alloc(size_t size);
```

Comment:
is this function public or used only in the wrapper macro?

Reply:
It is only here because it can be overwritten by the user.